### PR TITLE
Add flatten node + add shape to output node

### PR DESCRIPTION
**Pull request by stevenabreu7**
still need to add new tests for flatten

### CODE REVIEW

File: nir/ir.py [hunk 1]
```diff
@@ -120,6 +128,16 @@ class Delay(NIRNode):
     delay: np.ndarray  # Delay
 
 
+@dataclass
+class Flatten(NIRNode):
+    """Flatten node.
+
+    This node flattens the input tensor to a 1D tensor.
```
Comment (Jegp) [inline nir/ir.py:135]: Do we *always* want to flatten to a 1d tensor? Shouldn't this be specified somewhat like the PyTorch `Flatten` where you can select the dimensions to flatten?

For instance, something like `Flatten(start_dim, end_dim)`.

File: nir/ir.py [hunk 2]
```diff
@@ -237,7 +255,7 @@ class Output(NIRNode):
     Defines an output of the graph.
     """
 
-    pass
+    shape: int  # Size of output
```
Comment (Jegp) [inline nir/ir.py:258]: I believe the shape should be an `np.ndarray`

**Review by Jegp — CHANGES_REQUESTED** [review]
Neat additions! I have a few comments regarding the arguments of Input and the type of the `Output` shape parameter.

File: nir/ir.py (hunk 1 above)
Comment (matjobst) [reply nir/ir.py:135]: For me the current version seems not really consistent or it needs more documentation. In the docstring, it says that the Flatten should flatten to a 1D Tensor. But then there is a parameter n_dims, "Number of dimensions to flatten". Is this supposed to select which dimensions should be flattened? Then the docstring is wrong. Or I am misunderstanding the parameter.
But I also like the proposal from @Jegp .

**Comment by matjobst** [comment]
Maybe I missed the discussion. By my understanding the shape could be extracted from the node before the output node. For all other nodes (except Input, where it is necessary), we do not give the output shape. So why explicitly set the shape for the Output node?

**Comment by biphasic** [comment]
@matjobst has a good point, I also don't see the point of having to specify dimension of the output node

**Comment by Jegp** [comment]
> 

It'd be for convenience. Basically, a few people argued that it'd be helpful to directly read the shape directly from the graph instead of inferring it.

**Comment by stevenabreu7** [comment]
very true, just updated to follow pytorch convention

**Pull request merged**